src/workerd/jsg/modules-new-test.c++

```diff
@@ -0,0 +1,1889 @@
+      KJ_ASSERT(str == "Error: Module not found: foo:bar");
+    });
+
+    // Likewise, the bar module should be found when using a bundle context
+    js.tryCatch([&] {
+      auto value2 = ModuleRegistry::resolve(js, "file:///bar", "default"_kjc,
+                                            ResolveContext::Type::BUNDLE);
+      KJ_ASSERT(value2.isArrayBuffer());
+
+      // But should not be found from a built-in context
+      ModuleRegistry::resolve(js, "file:///bar", "default"_kjc,
+                              ResolveContext::Type::BUILTIN);
+      JSG_FAIL_REQUIRE(Error, "Should have thrown");
+    }, [&](Value exception) {
+      auto str = kj::str(exception.getHandle(js));
+      KJ_ASSERT(str == "Error: Module not found: file:///bar");
+    });
+  });
+}
+
+// ======================================================================================
+
+KJ_TEST("Awaiting top-level dynamic import in synchronous require fails as expected") {
+  PREAMBLE([&](Lock& js) {
+    ResolveObserverImpl observer;
+    CompilationObserver compilationObserver;
+
+    ModuleBundle::BundleBuilder bundleBuilder;
+    auto foo = kj::str("export default (await import('bar')).default;");
+    bundleBuilder.addEsmModule("foo", foo.slice(0, foo.size()).attach(kj::mv(foo)));
+
+    auto bar = kj::str("export default 123;");
+    bundleBuilder.addEsmModule("bar", bar.slice(0, bar.size()).attach(kj::mv(bar)));
+
+    auto registry = ModuleRegistry::Builder(observer).add(bundleBuilder.finish()).finish();
+
+    registry->attachToIsolate(js, compilationObserver);
+
+    js.tryCatch([&] {
+      // This is a synchronous resolve that should fail.
+      ModuleRegistry::resolve(js, "file:///foo", "default"_kjc);
+      KJ_FAIL_ASSERT("Should have failed");
+    }, [&](Value exception) {
+      auto str = kj::str(exception.getHandle(js));
+      KJ_ASSERT(str == "Error: The module evaluation did not complete synchronously. "
+                       "This is not permitted for synchronous require(...). "
+                       "Use await import(...) instead.");
+    });
+  });
+}
+
+// ======================================================================================
+
+KJ_TEST("Awaiting a never resolved promise in synchronous require fails as expected") {
+  PREAMBLE([&](Lock& js) {
+    ResolveObserverImpl observer;
+    CompilationObserver compilationObserver;
+
+    ModuleBundle::BundleBuilder bundleBuilder;
+    auto foo = kj::str("const p = new Promise(() => {}); await p;");
+    bundleBuilder.addEsmModule("foo", foo.slice(0, foo.size()).attach(kj::mv(foo)));
+
+    auto bar = kj::str("export default 123;");
+    bundleBuilder.addEsmModule("bar", bar.slice(0, bar.size()).attach(kj::mv(bar)));
+
+    auto registry = ModuleRegistry::Builder(observer).add(bundleBuilder.finish()).finish();
+
+    registry->attachToIsolate(js, compilationObserver);
+
+    js.tryCatch([&] {
+      // This is a synchronous resolve that should successfully wait on the async
```

Comment:
This comment is misleading and the bar variable above is redundant, bar is not being referenced anywhere.

Reply:
Copy'n'paste error